### PR TITLE
[Merged by Bors] - feat: the Unfolding Trick

### PR DESCRIPTION
**Pull request by AlexKontorovich**
This PR is a forward port of mathlib3 PR !3#18863

Co-authored-by: Heather Macbeth <25316162+hrmacbeth@users.noreply.github.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 1]
```diff
@@ -147,12 +146,169 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
...
+  {𝓕' : Set G'}
+
+local notation "μ_𝓕" => Measure.map (@QuotientAddGroup.mk G' _ Γ') (μ'.restrict 𝓕')
+
+/-- This is the **Unfolding Trick**: Given an additive subgroup `Γ'` of an additive group `G'`, the
+  integral of a function `f` on `G'` times the lift to `G'` of a function `g` on the quotient
+  `G' ⧸ Γ'` with respect to a right-invariant measure `μ` on `G'`, is equal to the integral over
+  the quotient of the automorphization of `f` times `g`. -/
+lemma QuotientAddGroup.integral_mul_eq_integral_automorphize_mul {K : Type _} [NormedField K]
+    [CompleteSpace K] [NormedSpace ℝ K] [μ'.IsAddRightInvariant] {f : G' → K}
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:282]:
```suggestion
    [NormedSpace ℝ K] [μ'.IsAddRightInvariant] {f : G' → K}
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 2]
```diff
@@ -147,12 +146,169 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
...
+      (measurePreserving_smul i μ).quasiMeasurePreserving).restrict
+  · rw [← h𝓕.lintegral_eq_tsum'' (fun x ↦ ‖f x‖₊)]
+    exact ne_of_lt hf₁.2
+
+/-- This is the **Unfolding Trick**: Given a subgroup `Γ` of a group `G`, the integral of a
+  function `f` on `G` times the lift to `G` of a function `g` on the quotient `G ⧸ Γ` with respect
+  to a right-invariant measure `μ` on `G`, is equal to the integral over the quotient of the
+  automorphization of `f` times `g`. -/
+lemma QuotientGroup.integral_mul_eq_integral_automorphize_mul {K : Type _} [NormedField K]
+    [CompleteSpace K] [NormedSpace ℝ K] [μ.IsMulRightInvariant] {f : G → K}
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:236]:
```suggestion
    [NormedSpace ℝ K] [μ.IsMulRightInvariant] {f : G → K}
```

File: Mathlib/GroupTheory/GroupAction/Group.lean [hunk 3]
```diff
@@ -265,6 +265,16 @@ def DistribMulAction.toAddAut : α →* AddAut β where
 #align distrib_mul_action.to_add_aut DistribMulAction.toAddAut
 #align distrib_mul_action.to_add_aut_apply DistribMulAction.toAddAut_apply
 
+/-- Each non-zero element of a `GroupWithZero` defines an additive monoid isomorphism of an
+`AddMonoid` on which it acts distributively.
+This is a stronger version of `DistribMulAction.toAddMonoidHom`. -/
+def DistribMulAction.toAddEquiv₀ {α : Type _} (β : Type _) [GroupWithZero α] [AddMonoid β]
```
Comment (ADedecker) [inline Mathlib/GroupTheory/GroupAction/Group.lean:271]:
```suggestion
@[simps!]
def DistribMulAction.toAddEquiv₀ {α : Type _} (β : Type _) [GroupWithZero α] [AddMonoid β]
```

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean [hunk 4]
```diff
@@ -1883,6 +1883,12 @@ theorem _root_.aestronglyMeasurable_withDensity_iff {E : Type _} [NormedAddCommG
     simpa only [NNReal.coe_eq_zero, Ne.def] using h'x
 #align ae_strongly_measurable_with_density_iff aestronglyMeasurable_withDensity_iff
 
+lemma aestronglyMeasurable_of_absolutelyContinuous {α β : Type _} [MeasurableSpace α]
```
Comment (ADedecker) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1886]: This is exactly [MeasureTheory.AEStronglyMeasurable.mono'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.html#MeasureTheory.AEStronglyMeasurable.mono') right?

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean [hunk 5]
```diff
@@ -253,6 +253,10 @@ theorem lintegral_eq_tsum' (h : IsFundamentalDomain G s μ) (f : α → ℝ≥0
 #align measure_theory.is_fundamental_domain.lintegral_eq_tsum' MeasureTheory.IsFundamentalDomain.lintegral_eq_tsum'
 #align measure_theory.is_add_fundamental_domain.lintegral_eq_tsum' MeasureTheory.IsAddFundamentalDomain.lintegral_eq_tsum'
 
+@[to_additive] lemma lintegral_eq_tsum'' (h : IsFundamentalDomain G s μ) (f : α → ℝ≥0∞) :
+  ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ :=
+(lintegral_eq_tsum' h f).trans ((Equiv.inv G).tsum_eq (fun g ↦ ∫⁻ (x : α) in s, f (g • x) ∂μ))
```
Comment (ADedecker) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]:
```suggestion
    ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ := by
  rw [← (Equiv.inv G).tsum_eq, lintegral_eq_tsum' h f]; rfl
```

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean [hunk 6]
```diff
@@ -426,6 +430,10 @@ theorem integral_eq_tsum' (h : IsFundamentalDomain G s μ) (f : α → E) (hf :
 #align measure_theory.is_fundamental_domain.integral_eq_tsum' MeasureTheory.IsFundamentalDomain.integral_eq_tsum'
 #align measure_theory.is_add_fundamental_domain.integral_eq_tsum' MeasureTheory.IsAddFundamentalDomain.integral_eq_tsum'
 
+@[to_additive] lemma integral_eq_tsum'' (h : IsFundamentalDomain G s μ)
+    (f : α → E) (hf : Integrable f μ) : ∫ x, f x ∂μ = ∑' g : G, ∫ x in s, f (g • x) ∂μ :=
+  (integral_eq_tsum' h f hf).trans ((Equiv.inv G).tsum_eq (fun g ↦ ∫ (x : α) in s, f (g • x) ∂μ))
```
Comment (ADedecker) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:435]:
```suggestion
    (f : α → E) (hf : Integrable f μ) : ∫ x, f x ∂μ = ∑' g : G, ∫ x in s, f (g • x) ∂μ := by
  rw [← (Equiv.inv G).tsum_eq, integral_eq_tsum' h f hf]; rfl
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 7]
```diff
@@ -113,16 +113,15 @@ theorem MeasureTheory.IsFundamentalDomain.isMulLeftInvariant_map [Subgroup.Norma
...
 /-- Given a normal subgroup `Γ` of a topological group `G` with Haar measure `μ`, which is also
   right-invariant, and a finite volume fundamental domain `𝓕`, the pushforward to the quotient
   group `G ⧸ Γ` of the restriction of `μ` to `𝓕` is a multiple of Haar measure on `G ⧸ Γ`. -/
 @[to_additive "Given a normal subgroup `Γ` of an additive topological group `G` with Haar measure
   `μ`, which is also right-invariant, and a finite volume fundamental domain `𝓕`, the pushforward
   to the quotient group `G ⧸ Γ` of the restriction of `μ` to `𝓕` is a multiple of Haar measure on
   `G ⧸ Γ`."]
-theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal Γ]
+theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient  [T2Space (G ⧸ Γ)]
+    [SecondCountableTopology (G ⧸ Γ)] (K : PositiveCompacts (G ⧸ Γ))[Subgroup.Normal Γ]
```
Comment (ADedecker) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:124]:
```suggestion
    [SecondCountableTopology (G ⧸ Γ)] (K : PositiveCompacts (G ⧸ Γ)) [Subgroup.Normal Γ]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 8]
```diff
@@ -147,12 +146,169 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
   topological group `G` with Haar measure `μ`, which is also right-invariant, and a finite volume
   fundamental domain `𝓕`, the quotient map to `G ⧸ Γ` is measure-preserving between appropriate
   multiples of Haar measure on `G` and `G ⧸ Γ`."]
-theorem MeasurePreservingQuotientGroup.mk' [Subgroup.Normal Γ]
-    [MeasureTheory.Measure.IsHaarMeasure μ] [μ.IsMulRightInvariant] (h𝓕_finite : μ 𝓕 < ⊤) (c : ℝ≥0)
+theorem MeasurePreservingQuotientGroup.mk' [T2Space (G ⧸ Γ)] [SecondCountableTopology (G ⧸ Γ)]
+    (K : PositiveCompacts (G ⧸ Γ))[Subgroup.Normal Γ] [MeasureTheory.Measure.IsHaarMeasure μ]
```
Comment (ADedecker) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:150]:
```suggestion
    (K : PositiveCompacts (G ⧸ Γ)) [Subgroup.Normal Γ] [μ.IsHaarMeasure]
```

**Comment by ADedecker** [comment]
Oh sorry, I had somehow missed the fact that this is a forward-port :facepalm: I'll delete some of my comments, the one left should still be applicable.

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 9]
```diff
@@ -147,12 +146,169 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
...
+  have meas_π : Measurable π := continuous_quotient_mk'.measurable
+  have H₀ : QuotientGroup.automorphize ((g ∘ π) * f) = g * (QuotientGroup.automorphize f) := by
+    exact QuotientGroup.automorphize_smul_left f g
+  calc ∫ (x : G), g (π x) * (f x) ∂μ =
+        ∫ (x : G ⧸ Γ), QuotientGroup.automorphize ((g ∘ π) * f) x ∂μ_𝓕 := ?_
+    _ = ∫ (x : G ⧸ Γ), g x * (QuotientGroup.automorphize f x) ∂μ_𝓕 := by simp [H₀]
+  have H₁ : Integrable ((g ∘ π) * f) μ
+  · have : AEStronglyMeasurable (fun (x : G) ↦ g (x : (G ⧸ Γ))) μ
+    · refine (AEStronglyMeasurable.aestronglyMeasurable_of_absolutelyContinuous ?_ _
+        hg).comp_measurable meas_π
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:252]:
```suggestion
    · refine (AEStronglyMeasurable.mono' hg ?_).comp_measurable meas_π
```

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean [hunk 10]
```diff
@@ -1883,6 +1883,12 @@ theorem _root_.aestronglyMeasurable_withDensity_iff {E : Type _} [NormedAddCommG
...
+    [TopologicalSpace β] {μ ν : Measure α} (h : ν ≪ μ) (g : α → β)
+    (hμ : AEStronglyMeasurable g μ) : AEStronglyMeasurable g ν := by
+  obtain ⟨g₁, hg₁, hg₁'⟩ := hμ
+  refine ⟨g₁, hg₁, h.ae_eq hg₁'⟩
+
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1891]:
```suggestion
```

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean (hunk 4 above)
Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean]: Yes, thanks! Deleted and updated.

**Comment by AlexKontorovich** [comment]
> 

Oh I see; I thought that was just for new files being moved, not files that were already ported but are now being modified. I tried clicking the link, but didn't follow which of that applied to this port? Which headers need to change, all of them?

**Comment by ADedecker** [comment]
Those of all files modified by your PR, yes. To find the right header, you should look on the [out-of-sync page](https://leanprover-community.github.io/mathlib-port-status/out-of-sync) and click on the file you are porting. Then the left column shows the changes in mathlib3 since the last synchronization. When you hover over he change you are porting, it highlights the corresponding change in the middle column (which is the output of mathport), and looking at the corresponding diff tells you what to change.

**Comment by AlexKontorovich** [comment]
> Those of all files modified by your PR, yes. To find the right header, you should look on the [out-of-sync page](https://leanprover-community.github.io/mathlib-port-status/out-of-sync) and click on the file you are porting. Then the left column shows the changes in mathlib3 since the last synchronization. When you hover over he change you are porting, it highlights the corresponding change in the middle column (which is the output of mathport), and looking at the corresponding diff tells you what to change.

Done, thanks!

**Review by YaelDillies — COMMENTED** [review]
This is a faithful forward-port, but where are all the `#align`s gone?

Whichever maintainer is reading this? Should we keep on adding the `#align`s? If so, please **do not merge this**.

maintainer merge

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 11]
```diff
@@ -147,12 +146,167 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
...
+  [Countable Γ'] [MeasurableSpace (G' ⧸ Γ')] [BorelSpace (G' ⧸ Γ')]
+  {𝓕' : Set G'}
+
+local notation "μ_𝓕" => Measure.map (@QuotientAddGroup.mk G' _ Γ') (μ'.restrict 𝓕')
+
+/-- This is the **Unfolding Trick**: Given an additive subgroup `Γ'` of an additive group `G'`, the
+  integral of a function `f` on `G'` times the lift to `G'` of a function `g` on the quotient
+  `G' ⧸ Γ'` with respect to a right-invariant measure `μ` on `G'`, is equal to the integral over
+  the quotient of the automorphization of `f` times `g`. -/
+lemma QuotientAddGroup.integral_mul_eq_integral_automorphize_mul {K : Type _} [NormedField K]
```
Comment (ericrbg) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:280]: what happens with `@[to_additive]` automatically here? should this be a bug report?

Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:280]: I think there are just too many things going on at once for `@[to_additive]` to work out which things should be additivized and which should remain the same...

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean (hunk 5 above)
Comment (sgouezel) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: indentation is wrong here

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 12]
```diff
@@ -1430,3 +1463,104 @@ theorem tsum_star : star (∑' b, f b) = ∑' b, star (f b) := by
...
+end
+
+section
+
+variable {G : Type _} [AddGroup G] {Γ : AddSubgroup G}
+
+/-- Automorphization of a function into an `R`-`module` distributes, that is, commutes with the
+`R`-scalar multiplication. -/
+lemma QuotientAddGroup.automorphize_smul_left (f : G → M) (g : G ⧸ Γ → R) :
+  QuotientAddGroup.automorphize ((g ∘ (@Quotient.mk' _ (_))) • f)
```
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1557]: indentation is wrong here.

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 13]
```diff
@@ -1287,10 +1287,43 @@ theorem Summable.const_smul (b : γ) (hf : Summable f) : Summable fun i => b •
...
 
+/-- Infinite sums commute with scalar multiplication. Version for scalars living in a `Monoid`, but
+  requiring a summability hypothesis. -/
 theorem tsum_const_smul [T2Space α] (b : γ) (hf : Summable f) : ∑' i, b • f i = b • ∑' i, f i :=
   (hf.hasSum.const_smul _).tsum_eq
 #align tsum_const_smul tsum_const_smul
 
+/-- Infinite sums commute with scalar multiplication. Version for scalars living in a `Group`, but
+  not requiring any summability hypothesis. -/
+lemma tsum_const_smul' {γ : Type _} [Group γ] [DistribMulAction γ α] [ContinuousConstSMul γ α]
```
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1298]: can you replace all `Type _` in the PR with `Type*`?

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean (hunk 5 above)
Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: sorry can you please tell me what's wrong? I don't see. Thanks!

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean (hunk 12 above)
Comment (AlexKontorovich) [reply Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1557]: Here too - what should the right indentation be?

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 14]
```diff
@@ -1430,3 +1463,104 @@ theorem tsum_star : star (∑' b, f b) = ∑' b, star (f b) := by
...
+lemma QuotientGroup.automorphize_smul_left (f : G → M) (g : G ⧸ Γ → R) :
+    (QuotientGroup.automorphize ((g ∘ (@Quotient.mk' _ (_)) : G → R) • f) : G ⧸ Γ → M)
+      = g • (QuotientGroup.automorphize f : G ⧸ Γ → M) :=
+  MulAction.automorphize_smul_left f g
+
+end
+
+section
+
+variable {G : Type _} [AddGroup G] {Γ : AddSubgroup G}
```
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1552]:
```suggestion
variable {G : Type*} [AddGroup G] {Γ : AddSubgroup G}
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 15]
```diff
@@ -1430,3 +1463,104 @@ theorem tsum_star : star (∑' b, f b) = ∑' b, star (f b) := by
...
+    use a
+  change ∑' a : α, g (π (a +ᵥ b)) • f (a +ᵥ b) = g (π b) • ∑' a : α, f (a +ᵥ b)
+  simp_rw [H₁]
+  exact tsum_const_smul'' _
+
+attribute [to_additive existing MulAction.automorphize_smul_left] AddAction.automorphize_smul_left
+
+section
+
+variable {G : Type _} [Group G] {Γ : Subgroup G}
```
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1532]:
```suggestion
variable {G : Type*} [Group G] {Γ : Subgroup G}
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 16]
```diff
@@ -1430,3 +1463,104 @@ theorem tsum_star : star (∑' b, f b) = ∑' b, star (f b) := by
 #align tsum_star tsum_star
 
 end ContinuousStar
+
+section automorphize
+
+variable {M : Type _} [TopologicalSpace M] [AddCommMonoid M] [T2Space M] {R : Type _}
```
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1469]:
```suggestion
variable {M : Type*} [TopologicalSpace M] [AddCommMonoid M] [T2Space M] {R : Type*}
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 17]
```diff
@@ -1287,10 +1287,43 @@ theorem Summable.const_smul (b : γ) (hf : Summable f) : Summable fun i => b •
...
+  apply tsum_eq_zero_of_not_summable
+  change ¬ Summable (mul_g ∘ f)
+  rwa [Summable.map_iff_of_equiv mul_g]
+  · apply continuous_const_smul
+  · apply continuous_const_smul
+
+/-- Infinite sums commute with scalar multiplication. Version for scalars living in a
+  `DivisionRing`; no summability hypothesis. This could be made to work for a
+  `[GroupWithZero γ]` if there was such a thing as `DistribMulActionWithZero`. -/
+lemma tsum_const_smul'' {γ : Type _} [DivisionRing γ] [Module γ α] [ContinuousConstSMul γ α]
```
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1314]:
```suggestion
lemma tsum_const_smul'' {γ : Type*} [DivisionRing γ] [Module γ α] [ContinuousConstSMul γ α]
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean (hunk 13 above)
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1298]:
```suggestion
lemma tsum_const_smul' {γ : Type*} [Group γ] [DistribMulAction γ α] [ContinuousConstSMul γ α]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean (hunk 11 above)
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:280]:
```suggestion
lemma QuotientAddGroup.integral_mul_eq_integral_automorphize_mul {K : Type*} [NormedField K]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 18]
```diff
@@ -147,12 +146,167 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
...
+  have H₂ : AEStronglyMeasurable (QuotientGroup.automorphize ((g ∘ π) * f)) μ_𝓕
+  · simp_rw [H₀]
+    exact hg.mul F_ae_measurable
+  apply QuotientGroup.integral_eq_integral_automorphize h𝓕 H₁ H₂
+
+end
+
+section
+
+variable {G' : Type _} [AddGroup G'] [MeasurableSpace G'] [TopologicalSpace G']
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:267]:
```suggestion
variable {G' : Type*} [AddGroup G'] [MeasurableSpace G'] [TopologicalSpace G']
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 19]
```diff
@@ -147,12 +146,167 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
...
+  · exact fun i ↦ (hf₁.1.comp_quasiMeasurePreserving
+      (measurePreserving_smul i μ).quasiMeasurePreserving).restrict
+  · rw [← h𝓕.lintegral_eq_tsum'' (fun x ↦ ‖f x‖₊)]
+    exact ne_of_lt hf₁.2
+
+/-- This is the **Unfolding Trick**: Given a subgroup `Γ` of a group `G`, the integral of a
+  function `f` on `G` times the lift to `G` of a function `g` on the quotient `G ⧸ Γ` with respect
+  to a right-invariant measure `μ` on `G`, is equal to the integral over the quotient of the
+  automorphization of `f` times `g`. -/
+lemma QuotientGroup.integral_mul_eq_integral_automorphize_mul {K : Type _} [NormedField K]
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:235]:
```suggestion
lemma QuotientGroup.integral_mul_eq_integral_automorphize_mul {K : Type*} [NormedField K]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 20]
```diff
@@ -147,12 +146,167 @@ theorem MeasureTheory.IsFundamentalDomain.map_restrict_quotient [Subgroup.Normal
...
+
+attribute [-instance] Quotient.instMeasurableSpace
+
+/-- This is a simple version of the **Unfolding Trick**: Given a subgroup `Γ` of a group `G`, the
+  integral of a function `f` on `G` with respect to a right-invariant measure `μ` is equal to the
+  integral over the quotient `G ⧸ Γ` of the automorphization of `f`. -/
+@[to_additive "This is a simple version of the **Unfolding Trick**: Given a subgroup `Γ` of an
+  additive  group `G`, the integral of a function `f` on `G` with respect to a right-invariant
+  measure `μ` is equal to the integral over the quotient `G ⧸ Γ` of the automorphization of `f`."]
+lemma QuotientGroup.integral_eq_integral_automorphize {E : Type _} [NormedAddCommGroup E]
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:216]:
```suggestion
lemma QuotientGroup.integral_eq_integral_automorphize {E : Type*} [NormedAddCommGroup E]
```

File: Mathlib/GroupTheory/GroupAction/Group.lean (hunk 3 above)
Comment (AlexKontorovich) [inline Mathlib/GroupTheory/GroupAction/Group.lean:271]:
```suggestion
def DistribMulAction.toAddEquiv₀ {α : Type*} (β : Type*) [GroupWithZero α] [AddMonoid β]
```

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean (hunk 5 above)
Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: Ok I think I've got it.

Comment (jcommelin) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: The indentation level of the type (aka statement) has changed with the switch to Lean 4.
The convention is now to indent by 4 spaces on a new line after the `:`.

Comment (jcommelin) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: If the statement is short, you can of course still write it on the same line as the `:` with only 1 space in between.

Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]:
```suggestion
    ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ := 
  (lintegral_eq_tsum' h f).trans ((Equiv.inv G).tsum_eq (fun g ↦ ∫⁻ (x : α) in s, f (g • x) ∂μ))
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 21]
```diff
@@ -1426,3 +1459,104 @@ theorem tsum_star : star (∑' b, f b) = ∑' b, star (f b) := by
...
+end
+
+section
+
+variable {G : Type*} [AddGroup G] {Γ : AddSubgroup G}
+
+/-- Automorphization of a function into an `R`-`module` distributes, that is, commutes with the
+`R`-scalar multiplication. -/
+lemma QuotientAddGroup.automorphize_smul_left (f : G → M) (g : G ⧸ Γ → R) :
+  QuotientAddGroup.automorphize ((g ∘ (@Quotient.mk' _ (_))) • f)
```
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1553]:
```suggestion
    QuotientAddGroup.automorphize ((g ∘ (@Quotient.mk' _ (_))) • f)
      = g • (QuotientAddGroup.automorphize f : G ⧸ Γ → M) :=
  AddAction.automorphize_smul_left f g
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean [hunk 22]
```diff
@@ -1426,3 +1459,104 @@ theorem tsum_star : star (∑' b, f b) = ∑' b, star (f b) := by
...
+    = g • (QuotientAddGroup.automorphize f : G ⧸ Γ → M) :=
+AddAction.automorphize_smul_left f g
```
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1555]:
```suggestion
    QuotientAddGroup.automorphize ((g ∘ (@Quotient.mk' _ (_))) • f)
      = g • (QuotientAddGroup.automorphize f : G ⧸ Γ → M) :=
  AddAction.automorphize_smul_left f g
```

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean [hunk 23]
```diff
@@ -253,6 +253,10 @@ theorem lintegral_eq_tsum' (h : IsFundamentalDomain G s μ) (f : α → ℝ≥0
 #align measure_theory.is_fundamental_domain.lintegral_eq_tsum' MeasureTheory.IsFundamentalDomain.lintegral_eq_tsum'
 #align measure_theory.is_add_fundamental_domain.lintegral_eq_tsum' MeasureTheory.IsAddFundamentalDomain.lintegral_eq_tsum'
 
+@[to_additive] lemma lintegral_eq_tsum'' (h : IsFundamentalDomain G s μ) (f : α → ℝ≥0∞) :
+    ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ :=
```
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:257]:
```suggestion
    ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ :=
```

**Comment by eric-wieser** [comment]
> Whichever maintainer is reading this? Should we keep on adding the `#align`s? If so, please **do note merge this**.

We should be keeping the aligns, because otherwise there wasn't much point merging the mathlib3 version.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by eric-wieser** [comment]
@jcommelin, I take it that's a statement that `#align`s don't matter any more?

**Comment by eric-wieser** [comment]
bors r-

(I think Johan's merge was in response to a maintainer merge comment that Yael deleted?)

**Comment by Ruben-VandeVelde** [comment]
> > Whichever maintainer is reading this? Should we keep on adding the `#align`s? If so, please **do note merge this**.
> 
> We should be keeping the aligns, because otherwise there wasn't much point merging the mathlib3 version.

I thought the main point of merging things into mathlib 3 at this point was to take advantage of mathport?

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**